README.md

```diff
@@ -248,6 +248,118 @@ public class ProductIndexSetUp : IndexMigration
+                }
+            }
+        }.ToBsonDocument();
+    }
+}
+```
+
+Document Migration:
+These are migrations performed on the documents of the collection to update the documents to add a new field or replace value of an existing field from the value of an existing field or some static value. You can also provide expression to apply some calculation/method(concat/sum/average) on the values before assigning it to the new field. The update is done via aggregation pipeline so you can also specify your own aggregation pipeline apart from $set/$addfield in the migration. You can also specify the filters on which the update should be applied. If not specified, it will be by default applied to all documents in the collection.   
+
+```csharp
+public class DocumentsUpdate_Revision4 : DocumentMigration
+{
+    public override Version Version => new Version(1, 1, 1, 4);
+    public override string Description => "documents update to new schema";
+    public override string CollectionName => "items";
+
+    public override void Prepare()
+    {
+        AddMigrationField("ProductDetails", "{ $concat: [ \"$Type\", \" - \", \"$ProductName\" ] }");
+        AddMigrationField("ProductType", "\"$Type\"");
+        AddMigrationField("Store.Region", "{ $concat: [ \"North \", \"$Store.Country\" ] }");
+            AddMigrationField("Sales", "{ $concatArrays: [ \"$Sales\", [ 55 ] ] }");
+            AddMigrationField("Ratings", "[ \"A\", \"B\" , \"$Rating\" ]");
+    }
+}
+
+public class DocumentsUpdate_Revision7 : DocumentMigration
+{
+    public override Version Version => new Version(1, 1, 1, 7);
+    public override string Description => "documents update to new schema";
+    public override string CollectionName => "items";
+    public override FilterDefinition<BsonDocument> Filters => BuildFilters();
+    public override PipelineDefinition<BsonDocument, BsonDocument> PipelineDefinition { get; set; }
+
+    public override void Prepare()
+    {
+        PipelineDefinition = BuildPipelineDefinition();
+    }
+
+    private static FilterDefinition<BsonDocument> BuildFilters()
+    {
+        var filterBuilder = new FilterDefinitionBuilder<BsonDocument>();
+        var idFilter = filterBuilder.Eq("ProductName", "Books");
+        var filter = filterBuilder.And(idFilter);
+        return filter;
+    }
+    
+    private static PipelineDefinition<BsonDocument, BsonDocument> BuildPipelineDefinition()
+    {
+        var pipeline = new EmptyPipelineDefinition<BsonDocument>()
+            .AppendStage("{ $addFields: { \"TargetGroup\": { \"$map\": { \"input\": \"$TargetGroup\", \"as\": \"row\", \"in\": { \"Buyer\": \"$$row.Buyer\", \"SellingPitch\": \"$$row.SellingPitch\"," + 
+                            " \"Genre\": \"$$row.SellingPitch\"  } } }}}",
+                BsonDocumentSerializer.Instance)
+            .AppendStage("{ $unset: \"Rating\" }", BsonDocumentSerializer.Instance);
+        return pipeline;
+    }
+}
+```
+
```

Comment:
We still need a document migration that is client side rather than server side - like with prism rights sets - I need to load the current rightset, restructure the data in c# and then save it back - for every single doc in the collection. In addition it needs to iterate while other database changes are happening and know when it is complete, so we have to use a find query to locate the next document that needs processing something like this:

Pseudocode
```
while(TryGetNextDocumentToProcess(out item))
{
   var migratedItem = MigrateDocument(item);
   collection.Save(migratedItem);
}
```

Reply:
I have split the DocumentMigration into ServerSide and ClientSide.